kafka-rest/src/test/java/io/confluent/kafkarest/v2/PartitionsResourceTest.java

```diff
@@ -0,0 +1,241 @@
+                new OffsetAndTimestamp(
+                    OFFSET_FOR_TIME, RESPONSE_TIMESTAMP_FOR_OFFSET.toEpochMilli())));
+    replay(adminClientWrapper, consumerManager);
+
+    String response =
+        target("/topics/{topic}/partitions/{partition}/offsets")
+            .resolveTemplate("topic", TOPIC)
+            .resolveTemplate("partition", PARTITION)
+            .queryParam("timestamp", DATE_TIME_FORMATTER.format(REQUEST_TIMESTAMP_FOR_OFFSET))
+            .request()
+            .get(String.class);
+
+    assertEquals(
+        String.format(
+            "{\"offset\":%d,\"timestamp\":\"%s\"}",
+            OFFSET_FOR_TIME, DATE_TIME_FORMATTER.format(RESPONSE_TIMESTAMP_FOR_OFFSET)),
+        response);
+
+    verify(adminClientWrapper, consumerManager);
+  }
+
+  @Test
+  public void getOffsets_withTimestamp_topicDoesNotExist_returns404() throws Exception {
+    expect(adminClientWrapper.topicExists(TOPIC)).andReturn(false);
+    replay(adminClientWrapper);
+
+    Response response =
+        target("/topics/{topic}/partitions/{partition}/offsets")
+            .resolveTemplate("topic", TOPIC)
+            .resolveTemplate("partition", PARTITION)
+            .queryParam("timestamp", DATE_TIME_FORMATTER.format(REQUEST_TIMESTAMP_FOR_OFFSET))
+            .request()
+            .get();
+
+    assertEquals(Status.NOT_FOUND.getStatusCode(), response.getStatus());
+
+    verify(adminClientWrapper);
+  }
+
+  @Test
+  public void getOffsets_withTimestamp_partitionDoesNotExist_returns404() throws Exception {
+    expect(adminClientWrapper.topicExists(TOPIC)).andReturn(true);
+    expect(adminClientWrapper.partitionExists(TOPIC, PARTITION)).andReturn(false);
+    replay(adminClientWrapper);
+
+    Response response =
+        target("/topics/{topic}/partitions/{partition}/offsets")
+            .resolveTemplate("topic", TOPIC)
+            .resolveTemplate("partition", PARTITION)
+            .queryParam("timestamp", DATE_TIME_FORMATTER.format(REQUEST_TIMESTAMP_FOR_OFFSET))
+            .request()
+            .get();
+
+    assertEquals(Status.NOT_FOUND.getStatusCode(), response.getStatus());
+
+    verify(adminClientWrapper);
+  }
+
+  @Test
+  public void getOffsets_withTimestamp_noSuchOffset_returnsEmpty() throws Exception {
+    expect(adminClientWrapper.topicExists(TOPIC)).andReturn(true);
+    expect(adminClientWrapper.partitionExists(TOPIC, PARTITION)).andReturn(true);
+    expect(consumerManager.getOffsetForTime(TOPIC, PARTITION, REQUEST_TIMESTAMP_FOR_OFFSET))
+        .andReturn(Optional.empty());
+    replay(adminClientWrapper, consumerManager);
+
+    String response =
+        target("/topics/{topic}/partitions/{partition}/offsets")
+            .resolveTemplate("topic", TOPIC)
+            .resolveTemplate("partition", PARTITION)
+            .queryParam("timestamp", DATE_TIME_FORMATTER.format(REQUEST_TIMESTAMP_FOR_OFFSET))
+            .request()
+            .get(String.class);
+
+    assertEquals("{}", response);
```

Comment:
this is an interesting case... I think `null` is semantically more correct. typically a query would return an array, and if nothing matched you'd return an empty array. here you return a value and if nothing matched you're returning a value of  a different form. as an alternative to using null, you could enclose the value in either an array or another object.

don't have strong conviction on the best thing to do here. do lean towards the `null` idea based on everything I can think of.

Reply:
Done.

Reply:
i mean `assertEquals("null", response)` - i think this is worse.